rclcpp/src/rclcpp/executors/executor_notify_waitable.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/executors/executor_notify_waitable.hpp"
+
+namespace rclcpp
+{
+namespace executors
+{
+
+ExecutorNotifyWaitable::ExecutorNotifyWaitable(std::function<void(void)> on_execute_callback)
+: execute_callback_(on_execute_callback)
+{
+}
+
+void
+ExecutorNotifyWaitable::add_to_wait_set(rcl_wait_set_t * wait_set)
+{
+  std::lock_guard<std::mutex> lock(guard_condition_mutex_);
+  for (auto guard_condition : this->notify_guard_conditions_) {
+    auto rcl_guard_condition = guard_condition->get_rcl_guard_condition();
+
+    rcl_ret_t ret = rcl_wait_set_add_guard_condition(
+      wait_set,
+      &rcl_guard_condition, NULL);
+
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(
+        ret, "failed to add guard condition to wait set");
+    }
+  }
+}
+
+bool
+ExecutorNotifyWaitable::is_ready(rcl_wait_set_t * wait_set)
+{
+  std::lock_guard<std::mutex> lock(guard_condition_mutex_);
+  for (size_t ii = 0; ii < wait_set->size_of_guard_conditions; ++ii) {
+    auto rcl_guard_condition = wait_set->guard_conditions[ii];
+
+    if (nullptr == rcl_guard_condition) {
+      continue;
+    }
+
+    for (auto guard_condition : this->notify_guard_conditions_) {
+      if (&guard_condition->get_rcl_guard_condition() == rcl_guard_condition) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void
+ExecutorNotifyWaitable::execute(std::shared_ptr<void> & data)
+{
+  (void) data;
+  this->execute_callback_();
+}
+
+std::shared_ptr<void>
+ExecutorNotifyWaitable::take_data()
+{
+  return nullptr;
+}
+
+void
+ExecutorNotifyWaitable::add_guard_condition(const rclcpp::GuardCondition * guard_condition)
+{
+  if (guard_condition == nullptr) {
+    throw std::runtime_error("Attempting to add null guard condition.");
+  }
+  std::lock_guard<std::mutex> lock(guard_condition_mutex_);
+  to_add_.push_back(guard_condition);
+}
+
+void
+ExecutorNotifyWaitable::remove_guard_condition(const rclcpp::GuardCondition * guard_condition)
+{
+  if (guard_condition == nullptr) {
+    throw std::runtime_error("Attempting to remove null guard condition.");
+  }
+  std::lock_guard<std::mutex> lock(guard_condition_mutex_);
+  to_remove_.push_back(guard_condition);
+}
+
+size_t
+ExecutorNotifyWaitable::get_number_of_ready_guard_conditions()
```

Comment:
I'm not sure I understand this.
Why is the `get_number_of_ready_guard_conditions` function responsible for updating the `notify_guard_conditions_` list ?

What's the advantage of having the 3 lists `to_add`, `to_remove` and `notify_guard_conditions_` ?

Reply:
The three lists are to prevent the `notify_guard_conditions_` from being mutated while being waited on in the wait set.

I placed it in the `get_number_of_ready_guard_conditions` because when the waitable is used by the executor, the following happens:

* `get_number_of_ready_guard_conditions` - to check how many conditions to add from this waitable, and is accounted for when resizing the waitset
* `add_to_wait_set` - allows the waitable to insert itself into the waitset (but no more than what was returned from `get_number_of_ready_guard_conditions`
* `is_ready` - allows the waitable to check it's entries in the waitset, here is where we need the corresponding list of guard conditions to check against
* `take_data` and `execute` - runs the contents of the user callback

So in this case, if `add_node` or `remove_node` are called from another thread while the executor is spinning, we hold those guard conditions in a temporary location to be added when the next update comes.

Reply:
The other option would be to queue node/callback_group updates and process those when we re-collect entities?

Reply:
Ok.
The events-executor needs to register a callback when a guard condition is added and remove it when it is removed.
Outside of that, it doesn;t need to hold on the callbacks.